File: Editorial/59_Spiral_Matrix_II/Go/2_Optimized_spiral_traversal.go
func generateMatrix(n int) [][]int {
    result := make([][]int, n)
    for i := range result {
        result[i] = make([]int, n)
    }
    dirs := [4][2]int{{0, 1}, {1, 0}, {0, -1}, {-1, 0}}
    d := 0
    row := 0
    col := 0
    cnt := 1
    for cnt <= n*n {
        result[row][col] = cnt
        cnt++
        newRow := (row + dirs[d][0] + n) % n
        newCol := (col + dirs[d][1] + n) % n
        // change direction if next cell is non zero
        if result[newRow][newCol] != 0 {
            d = (d + 1) % 4
        }
        row += dirs[d][0]
        col += dirs[d][1]
    }
    return result
}

File: Editorial/0059_Spiral_Matrix_II/JavaScript/2_Optimized_spiral_traversal.js
var generateMatrix = function (n) {
    const result = new Array(n).fill(0).map(() => new Array(n).fill(0));
    const dirs = [
        [0, 1],
        [1, 0],
        [0, -1],
        [-1, 0],
    ];
    let d = 0;
    let row = 0;
    let col = 0;
    let cnt = 1;
    while (cnt <= n * n) {
        result[row][col] = cnt++;
        let newRow = (row + (dirs[d][0] % n) + n) % n;
        let newCol = (col + (dirs[d][1] % n) + n) % n;
        //change direction if next cell is non zero
        if (result[newRow][newCol] != 0) d = (d + 1) % 4;
        row += dirs[d][0];
        col += dirs[d][1];
    }
    return result;
};
